tests/unitTests/src/unitTestsStorage.ts

```diff
@@ -737,6 +737,58 @@ export function runStorageTests() {
       })
   })
 
+  test('putFile passes etag to upload function', async (t) => {
```

Comment:
The `spy` test pattern here is cool. Nice improvement over the way the similar tests work.

I think we need at least one more test, like this one, that triggers the new PreconditionFailed error path.